### PR TITLE
[CHORE] Project clean up

### PR DESCRIPTION
**Pull request by regniblod**
I decided to start contributing in this project since I need to implement some new features for an internal trivago project that uses this project.

To start contributing I (and everyone else who wants to contribute as well) need a clean starting point, so I decided to do a bit of clean up of the project.

This is what I changed:
    * Change company name in license from `trivago` to `trivago GmbH` (the official name).
    * Change license year from `2017` to `2017-present` (as everyone does in the other open source trivago projects).
    * Added DocBlock for every method that needed one, specifying all types of arguments and return types, along with all the possible thrown exceptions.
    * Applied coding style where needed.
    * Removed unused attributes in some classes.
    * Fixed small typo in the README.

### CODE REVIEW

File: Trivago/Jade/Application/Framework/JadeBundle/Controller/JsonApiController.php [hunk 1]
```diff
@@ -183,9 +196,17 @@ public function createEntityAction($resourceName, Request $httpRequest)
...
-     * @param string $resourceName
+     * @param string  $resourceName
      * @param Request $httpRequest
+     *
      * @return JsonApiResponse
+     *
+     * @throws InvalidPath
+     * @throws AccessDeniedException
+     * @throws InvalidRequest
+     * @throws \LogicException
```
Comment (moein) [inline Trivago/Jade/Application/Framework/JadeBundle/Controller/JsonApiController.php]: Normally you only define exceptions that are meant to be captured specially for services.
I don't really see a point of defining this in the controllers
And it wasn't a controller \LogicException is not to be caught at all!

File: Trivago/Jade/Application/Framework/JadeBundle/DependencyInjection/Compiler/ListenerPass.php [hunk 2]
```diff
@@ -36,7 +35,11 @@ class ListenerPass implements CompilerPassInterface
     const TAG = 'trivago_jade.listener';
 
     /**
-     * @inheritdoc
+     * {@inheritdoc}
+     *
+     * @throws InvalidArgumentException
```
Comment (moein) [inline Trivago/Jade/Application/Framework/JadeBundle/DependencyInjection/Compiler/ListenerPass.php:40]: Again not really needed as it gives no useful information

File: Trivago/Jade/Application/Framework/JadeBundle/DependencyInjection/TrivagoJadeExtension.php [hunk 3]
```diff
@@ -42,6 +39,8 @@ class TrivagoJadeExtension extends Extension
 {
     /**
      * {@inheritdoc}
+     *
+     * @throws \Exception
```
Comment (moein) [inline Trivago/Jade/Application/Framework/JadeBundle/DependencyInjection/TrivagoJadeExtension.php:43]: Not needed

File: Trivago/Jade/Application/JsonApi/Mapping/RuntimeResourceMapper.php [hunk 4]
```diff
@@ -95,11 +102,10 @@ private function setupAttributes(\ReflectionClass $reflection, ResourceConfig $r
                 continue;
             }
 
-            if ($resourceConfig->hasAttributePermissionFor($propertyName)) {
-                if (!$this->hasPermission($resource, $resourceConfig->getAttributePermissionsFor($propertyName))) {
+            if ($resourceConfig->hasAttributePermissionFor($propertyName)
+                && !$this->hasPermission($resource, $resourceConfig->getAttributePermissionsFor($propertyName))) {
                     continue;
                 }
```
Comment (moein) [inline Trivago/Jade/Application/JsonApi/Mapping/RuntimeResourceMapper.php]: Wrong intend on this line

File: Trivago/Jade/Application/JsonApi/Request/ActionRequest.php [hunk 5]
```diff
@@ -14,16 +14,24 @@
...
  */
 
 namespace Trivago\Jade\Application\JsonApi\Request;
 
 class ActionRequest extends ResourceRequest
 {
+    /**
+     * @var mixed
+     */
     protected $id;
```
Comment (moein) [inline Trivago/Jade/Application/JsonApi/Request/ActionRequest.php]: id in json api is always a string

File: Trivago/Jade/Application/JsonApi/Request/DeleteRequest.php [hunk 6]
```diff
@@ -30,7 +27,7 @@ class DeleteRequest extends ResourceRequest
 
     /**
      * @param string $resourceName
-     * @param mixed $id
+     * @param mixed  $id
```
Comment (moein) [inline Trivago/Jade/Application/JsonApi/Request/DeleteRequest.php]: string

File: Trivago/Jade/Application/JsonApi/Request/EntityRequest.php [hunk 7]
```diff
@@ -39,8 +36,8 @@ class EntityRequest extends ResourceRequest
     protected $relationships;
 
     /**
-     * @param string $resourceName
-     * @param mixed $id
+     * @param string   $resourceName
+     * @param mixed    $id
```
Comment (moein) [inline Trivago/Jade/Application/JsonApi/Request/EntityRequest.php]: string

File: Trivago/Jade/Application/JsonApi/Request/RequestFactory.php [hunk 8]
```diff
@@ -151,7 +161,8 @@ public function createEntityRequest(ParameterBag $query, $resourceName, $resourc
 
     /**
      * @param string $resourceName
-     * @param mixed $resourceId
+     * @param mixed  $resourceId
```
Comment (moein) [inline Trivago/Jade/Application/JsonApi/Request/RequestFactory.php]: string

File: Trivago/Jade/Application/JsonApi/Request/UpdateRequest.php [hunk 9]
```diff
@@ -32,13 +29,17 @@ class UpdateRequest extends CreateRequest
     protected $id;
 
     /**
-     * @param string $resourceName
-     * @param mixed $id
-     * @param ResourceAttributeBag $attributes
+     * @param string                  $resourceName
+     * @param mixed                   $id
```
Comment (moein) [inline Trivago/Jade/Application/JsonApi/Request/UpdateRequest.php]: string

File: Trivago/Jade/Application/JsonApi/Schema/InvalidRequest.php [hunk 10]
```diff
@@ -28,30 +25,47 @@ class InvalidRequest extends \RuntimeException
...
+     *
      * @return InvalidRequest
      */
     public static function createWithMessage($id, $code, $message)
     {
         return new self([new Error($id, null, null, $code, $message)]);
     }
 
+    /**
+     * @param mixed  $id
```
Comment (moein) [inline Trivago/Jade/Application/JsonApi/Schema/InvalidRequest.php]: string

**Review by moein — COMMENTED** [review]
Really good job
Please make sure all resource ids are string
And remove all the throws unless in service methods that the exceptions are meant to be captured

**Comment by regniblod** [comment]
Hey @moein, I fixed all your comments except for the Exception ones.

I know what you mean with "they don't provide useful information" but maybe that means we have to give them meaningful information?

I was thinking about converting all the `\InvalidArgumentException` and such into meaningful exceptions from the Domain.

Otherwise what's the value on throwing an exception that I'm not even supposed to see in the DocBlock from the project using JADE? I'm supposed to just wrap every single call to a JADE method into a `try { ... } catch (\Exception $e) {}` just in case?

**Comment by moein** [comment]
@regniblod Exceptions like \InvalidArgumentException or \LogicException are not meant to be captured at all.
These are exceptions that developers should see in the dev testing. For example, if you have a method that receives 2 numbers: father's age and son's age. If any of these ages is not a positive integer it's an InvalidArgumentException. In this case, the person who is calling the method has to take care of this constraint.
Also, the father's age has to be greater than son's age so if that's not the case it's a LogicException since each argument by itself is valid but not the conjunction.
And finally if you have a constraint from the business side that the father has to be older than 18 then that's when you use domain exception and that's the only exception the developer who is calling the method is supposed to catch as it might change in the future.
I hope it's clear enough

**Comment by regniblod** [comment]
Thanks for the explanation @moein!

I did 2 more commits, I hope everything looks better now.

**Pull request merged**